src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,238 @@
+        //
+        // It's ok to come up with a file-enumeration that includes too much as the glob filter
+        // will filter out the extra, but it's not ok to include too little in the enumeration.
+        internal static void DetermineFileEnumerationFromGlob(
+            string includeGlobPathAbsolute,
+            out string enumerateRootPath,
+            out string enumeratePattern,
+            out SearchOption enumerateDepth)
+        {
+            int firstGlob = includeGlobPathAbsolute.IndexOfAny(GlobChars);
+            bool hasRecursive = includeGlobPathAbsolute.Contains("**", StringComparison.Ordinal);
+
+            // no globbing
+            if (firstGlob < 0)
+            {
+                enumerateRootPath = Path.GetDirectoryName(includeGlobPathAbsolute);
+                enumeratePattern = Path.GetFileName(includeGlobPathAbsolute);
+                enumerateDepth = SearchOption.TopDirectoryOnly;
+            }
+            else
+            {
+                int rootDirLength = includeGlobPathAbsolute.Substring(0,firstGlob).LastIndexOfAny( new [] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar});
+                enumerateRootPath = includeGlobPathAbsolute.Substring(0,rootDirLength);
+                enumeratePattern = "*";
+                enumerateDepth = hasRecursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
+            }
+        }
+
+        public static Fingerprint EvaluateKeyToFingerprint(
+            AgentTaskPluginExecutionContext context,
+            string filePathRoot,
+            IEnumerable<string> keySegments,
+            bool addWildcard)
+        {
+            var sha256 = new SHA256Managed();
+
+            string workingDirectoryValue = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                if (keySegment.Equals("*", StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`*` is a reserved key segment. For path glob, use `./*`.");
+                }
+                else if (keySegment.Equals(Fingerprint.Wildcard, StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`**` is a reserved key segment. For path glob, use `./**`.");
+                }
+                else if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    var segment = new StringBuilder();
+                    bool foundFile = false;
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries);
+                    string rootRule = pathRules.First();
+                    if(rootRule[0] == '!')
```

Comment:
nit: to be consistent with above, consider using `First()` instead of `[0]`. Might also just consider using `StartsWith` and `EndsWith` everywhere.